src/WebJobs.Script.Abstractions/Description/IFunctionProvider.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.Azure.WebJobs.Script.Description;
+
+namespace Microsoft.Azure.WebJobs.Script.Abstractions
+{
+    public interface IFunctionProvider
+    {
+        /// <summary>
+        /// Gets any function errors that may occur as part of the provider context
+        /// </summary>
+        /// <returns> An IDictionary of function name to the list of errors</returns>
+        IDictionary<string, ICollection<string>> GetFunctionErrors();
+
+        /// <summary>
+        /// Gets all function metadata that this provider knows about
+        /// </summary>
+        /// <returns>An IEnumerable of FunctionMetadata</returns>
+        IEnumerable<FunctionMetadata> GetFunctionMetadata();
```

Comment:
It was requested that this be made `async` (though this is going to be a blocking call) so in future, extensions can take advantage of async if that happens, without restructuring their code.

Reply:
Yes, making this async would be good